MRG: Automatic artifact detection with FASTER

wmvanvliet opened the pull request:
The FASTER method consists of a series of algorithms designed to automatically detect various artifacts through outlier detection [1]. This PR implements ~~the first four~~ some steps of the series:

1. Detect bad MEG/EEG channels
2. Detect bad epochs
3. ~~Detect bad ICA components~~
4. Detect for each epoch, channels that are locally bad

An example is provided that takes the user through all four steps and demonstrates how to deal with the detected artifacts (bad channels can be interpolated, bad epochs can be rejected, ~~bad ICA components can be zeroed out~~, ...)

[1] Nolan H., Whelan R. and Reilly RB. FASTER: fully automated statistical thresholding for EEG artifact rejection. Journal of Neuroscience Methods, vol. 192, issue 1, pp. 152-162, 2010. http://doi.org/10.1016/j.jneumeth.2010.07.015

wmvanvliet commented:
TODO:
- [ ] Write unit tests
- [ ] Extensive testing on multiple datasets
- [ ] Come up with a user friendly API for the correction of the detected artifacts

wmvanvliet commented:
@dengemann any thoughts?

dengemann commented on examples/preprocessing/plot_artifact_correction_with_faster.py:
```diff
@@ -0,0 +1,119 @@
+# Compute evoked before cleaning, using an average EEG reference
+epochs_before = epochs.copy()
+epochs_before.info['custom_ref_applied'] = False
+epochs_before, _ = io.set_eeg_reference(epochs_before)
```
btw. I still think this is weird to have in io

wmvanvliet replied:
yes. Feel free to move it :)

wmvanvliet replied:
attempted to move it... created circular import... gave up.
Also, average referencing is handled in `io.proj`, so there are some links to other pieces of the `io` module.

dengemann commented on examples/preprocessing/plot_artifact_correction_with_faster.py:
```diff
@@ -0,0 +1,119 @@
+
+# Reject bad epochs (keep good epochs)
+good_epochs = list(set(range(len(epochs))).difference(set(bad_epochs)))
+epochs = epochs[good_epochs]
```
this is complicated, why not use `Epochs.drop_bad_epochs`?

wmvanvliet replied:
I looked into that function... but it turned out to reject epochs based on the `is_good` function that performs peak to peak detection.... no way for me to easily use it to reject a known list of bad epochs.

dengemann commented on examples/preprocessing/plot_artifact_correction_with_faster.py:
```diff
@@ -0,0 +1,119 @@
+bad_epochs = faster_bad_epochs(epochs, picks=eeg_picks, thres=3)
+
+meg_picks = mne.pick_types(epochs.info, meg=True, exclude='bads')
+bad_epochs += faster_bad_epochs(epochs, picks=meg_picks, thres=3)
```
the API is complicated. should do multi channel handlin in one call

wmvanvliet replied:
yes it should.

dengemann commented on examples/preprocessing/plot_artifact_correction_with_faster.py:
```diff
@@ -0,0 +1,119 @@
+ecg_epochs = create_ecg_epochs(raw, picks=picks)
+eog_components, _ = ica.find_bads_eog(eog_epochs)
+ecg_components, _ = ica.find_bads_ecg(ecg_epochs)
+ica.exclude = eog_components + ecg_components
```
ICA faster is still missin here, right?

wmvanvliet replied:
It's sort of implemented as `faster_bad_components`, but I have a feeling the existing ICA artifact detection/rejection procedures are superior to FASTER.

dengemann replied:
but only if you've got ECG + EOG channels

dengemann commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,368 @@
+    y = np.cumsum(np.diff(x, axis=1), axis=1)
+
+    b1 = [1, -2, 1]
+    b2 = [1,  0, -2, 0, 1]
```
white spaces

dengemann commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,368 @@
+    fs, ps = _efficient_welch(data, sfreq)
+
+    # Limit power spectrum to upper frequencies
+    ps = ps[:, np.searchsorted(fs, 25):np.searchsorted(fs, 45)]
```
why is this hardcoded?

wmvanvliet replied:
because I'm lazy. This part needs more testing and looking in the original paper / EEGLAB code to find the exact procedure.

dengemann commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,368 @@
+        logger.info('Bad by %s:\n\t%s' % (m, b))
+        bads.append(b)
+
+    return np.unique(np.concatenate(bads)).tolist()
```
for plotting purposes it would be cool to have an option that returns a dict of metric_name: result

wmvanvliet replied:
but for regular use, you want a plain list. The bads detected by each method is already listed in the log... would plotting the bads detected by each method be a frequent enough use case to warrant an extra parameter?

dengemann replied:
.... yes for me ... we can have a bool flag for expert mode. return_metrics=False

dengemann commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,368 @@
+    bads = []
+    for m in use_metrics:
+        s = metrics[m](data)
+        b = find_outliers(s, thres)
```
for the real re-implementation of faster you need to use the max_iter parameter with 1 here and we should expose it to allow for different values

dengemann commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,368 @@
+    for m in use_metrics:
+        scores = np.atleast_2d(metrics[m](ica))
+        for s in scores:
+            b = find_outliers(s, thres)
```
same here ...

dengemann commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,368 @@
+    ICA.find_bads_ecg
+    ICA.find_bads_eog
+    """
+    source_data = ica.get_sources(epochs).get_data().transpose(1,0,2)
```
whitespace missing

dengemann commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,368 @@
+
+    for i, b in enumerate(bads):
+        if len(b) > 0:
+            bads[i] = np.unique(np.concatenate(b)).tolist()
```
if bads is a list we don't need to enumerate but can just use append

dengemann replied:
looks like a filter list comprehension could do it much easier:

`bads =  [np.unique(np.concatenate(b)).tolist() for b in bads if len(b) >0]`

wmvanvliet replied:
yes, that looks better.

dengemann commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,368 @@
+    if use_metrics is None:
+        use_metrics = metrics.keys()
+
+    bads = []
```
let's use `list()` here instead of the literal, it's a convention we follow because it's easier to get for newcomers and more explicit

dengemann commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,368 @@
+    
+    data = epochs.get_data()[:, picks, :]
+
+    bads = [[] for i in range(len(epochs))]
```
`[[] for _ in range(len(epochs))]`

dengemann commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,368 @@
+        'kurtosis':        lambda x: kurtosis(
+                               np.dot(
+                                   x.mixing_matrix_.T,
+                                   x.pca_components_[:x.n_components_]),
```
the whitenind / desphering step is missing ... see `ica._pre_whitener`

dengemann commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,368 @@
+        'line_noise':      lambda x: _freqs_power(source_data,
+                                                  epochs.info['sfreq'],
+                                                  [50, 60]),
+    }
```
btw there are a couple of pep8 errors, e.g. see white space after keys in the dicts

wmvanvliet replied:
lets worry about PEP8 later, when we are happy with the functionality.

dengemann commented:
@wmvanvliet where are we here? Hagnout soon?

wmvanvliet commented:
I'm currently figuratively working myself to death trying to finish my dissertation and secure a postdoc grant. When those are submitted I'll look into FASTER again. In the meanwhile, please feel free to fork my branch!

The API needs love... you made some excellent suggestions.
And it needs testing! For now, I've tested it on 5 EEG datasets I was working on at the time and it performed very well. It would be good to set up something more substantial.

wmvanvliet commented:
@dengemann I've made some tweaks to the API to make it nicer. See the updated example script. Could you take a look at the `faster_bad_components` function? I have too little knowledge of the ICA code to make it work properly.

wmvanvliet commented:
We should start thinking about how to evaluate FASTER.
In the paper, part of their evaluation is on artificial data.
The evaluation on real data is performed on an P300 oddball dataset.
Performance measures were the variation of the signal during the baseline interval (-0.2 -- 0) (smaller is better) and the size of the P300 in the ERP (bigger is better, they don't specify how they measured this exactly).

wmvanvliet commented:
(rebased)

Eric89GXL commented:
Is this still WIP, or ready to go from your end? I'm assuming we won't do the performance evaluation as part of this PR...?

There are currently no tests. One thing I would do, partially as an evaluation and also to get tests, is implement some tests that you know it should pass if it's implemented correctly. Whatever types of artifacts it should find, create those using RawArray or EpochsArray, and make sure it works. WDYT?

Eric89GXL commented:
@agramfort I'd like to re-enable coveralls comments. They make the comment string longer, but make it clearer when we should be concerned about coverage. Okay with you?

dengemann commented:
+1

2015-04-17 18:02 GMT+02:00 Eric Larson <notifications@github.com>:

> @agramfort <https://github.com/agramfort> I'd like to re-enable coveralls
> comments. They make the comment string longer, but make it clearer when we
> should be concerned about coverage. Okay with you?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1777#issuecomment-94016291>.
>

agramfort commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,414 @@
+import numpy as np
```
there is not author or license header

agramfort commented:
this PR addresses the generic problem of automatic artifact rejection. How
close are we to have a find_bad_channels and find_bad_epochs? I'd like to
avoid having different functions with different APIs that address the same
problem.

dengemann commented:
I think we need more time for this. It does not make sense to rush here. It needs some testing.

dengemann commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,414 @@
+    for ch in picks:
+        ch_types[channel_type(info, ch)].append(ch)
+
+    return ch_types.items()
```
.... we have such a function already somewhere in picks, even two of them, one public, one private.

dengemann commented:
Also the API is not yet fully settled I feel.

agramfort commented:
>
> I think we need more time for this. It does not make sense to rush here.
> It needs some testing.
>
+1

jona-sassenhagen commented:
Don't other functions use 'threshold' for their keyword (not 'thresh')?

Also, I get this:
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-15-46491baf3dc6> in <module>()
     21 
     22 if True:
---> 23     epochs.info['bads'] = faster_bad_channels(epochs, thres=5, max_iter=1)
     24     raw.info["bads"] = deepcopy(epochs.info['bads'])
     25     if len(epochs.info['bads']) > 0:

/home/FB/fb05/sassenha/tools/mne-working/mne-python/mne/preprocessing/faster.py in faster_bad_channels(epochs, picks, thres, use_metrics, max_iter)
    204         logger.info('Bad channel detection on %s channels:' % ch_type.upper())
    205         for m in use_metrics:
--> 206             s = metrics[m](data[chs])
    207             b = [epochs.ch_names[picks[chs[i]]]
    208                  for i in find_outliers(s, thres, max_iter)]

/home/FB/fb05/sassenha/tools/mne-working/mne-python/mne/preprocessing/faster.py in <lambda>(x)
    186         'kurtosis': lambda x: kurtosis(x, axis=1),
    187         'line_noise': lambda x: _freqs_power(x, epochs.info['sfreq'],
--> 188                                              [50, 60]),
    189     }
    190 

/home/FB/fb05/sassenha/tools/mne-working/mne-python/mne/preprocessing/faster.py in _freqs_power(data, sfreq, freqs)
     97     """
     98     fs, ps = _efficient_welch(data, sfreq)
---> 99     return np.sum([ps[..., np.searchsorted(fs, f)] for f in freqs], axis=0)
    100 
    101 

/home/FB/fb05/sassenha/tools/mne-working/mne-python/mne/preprocessing/faster.py in <listcomp>(.0)
     97     """
     98     fs, ps = _efficient_welch(data, sfreq)
---> 99     return np.sum([ps[..., np.searchsorted(fs, f)] for f in freqs], axis=0)
    100 
    101 

IndexError: index 513 is out of bounds for axis 1 with size 513
```

The others seem to work.

wmvanvliet commented:
> How close are we to have a find_bad_channels and find_bad_epochs?

I'm using them in my own pipelines and are quite happy with their performance.

I'll look into @jona-sassenhagen's bug. It seems that the 50 and 60 Hz frequencies are not present in the spectrum that was calculated. This is never checked for.

At the top of the PR, there is a check list. You can see there that this PR needs unit tests, some sort of performance testing, and API work.

I'm in to particular hurry to merge this one quickly.

wmvanvliet commented:
Does anyone have any idea how to test performance on this thing?
Of course creating an artificial dataset with some generated noise is good for unit testing, but I would like to have some indication how well this performs 'in the wild' as well.

What would be a good metric to use?

dengemann commented:
@fraimondo and I plan to test this for example against a standard EGI system based preprocessing workflow on rouhgly 200 EEG datasets.

jona-sassenhagen commented:
@wmvanvliet you're correct, I think the specific file was downsampled to 100 Hz. Maybe a warning, possibly automatically turning off that metric ...

jona-sassenhagen commented:
In the current implementation, when running faster_bad_channels with return_by_metric=False, I get an error that the input to np.concatenate must be a sequence. It works if you replace the return of the figure with this: ```return np.unique(np.concatenate(list(bads.values()))).tolist()```

dengemann commented:
@wmvanvliet 
@jona-sassenhagen 

@fraimondo and I are ready -- 329 EEG recordings are waiting for getting FASTERed.
Maybe we could hangout next Tuesday to think about a proper validation protocol?
Please get back to us.

jona-sassenhagen commented:
Another one: `reduce` (line 455) has been moved to functools in Python 3.

wmvanvliet commented:
(rebased)

wmvanvliet commented:
@jona-sassenhagen your comments have been addressed (thanks!)

 1. When you downsample the signal too much, some helpful warning messages are now presented if some metrics are used. For example, 50 Hz power line noise detection cannot work with a sample rate of ≤100 Hz. Same for the 'power gradient' metric.
 2. Python3 no longer returns dict values as a list... so they are now wrapped in `list()`
 3. Python3 moved `reduce`... so the proper import line is now added.
 4. Rebased the entire thing and resolved a merge conflict.

wmvanvliet commented:
@dengemann could you look at [line 72 of the example script](https://github.com/mne-tools/mne-python/commit/d91cc67aeac912ed08297eb15f708baf3edf5cb8)? When I set `eog=True` there it bugs out. Maybe I'm calling ICA in the wrong fashion or there's a bug in the ICA code.

dengemann commented:
@wmvanvliet  will have a look asap.
@jona-sassenhagen @wmvanvliet should we hang out soon with @fraimondo to discuss a validation protocol?

wmvanvliet commented:
To quote the paper:

> In order to quantify the performance of the artifact removal
method on real data, a metric for measuring artifact contribution to EEG was defined as the variance in the 500ms baseline prior to the stimulus presentation (Handy et al., 2003). This variance should be lower in datasets with fewer artifacts. This metric has the advantage of making no assumptions about the signal content. The disadvantage is that a signal with lower amplitude has lower variance, and therefore any method that removed all data would have perfect performance. Therefore, P3 amplitude was also reported, in order to demonstrate that the EEG signal was not being attenuated. The percentage of epochs removed was also recorded, as a measure of data retention. These measures were compared for each processing method.

It doesn't say how P3 amplitude was measured.

wmvanvliet commented:
I propose the following metric: we extract epochs, perform FASTER, then compute for each time sample, the standard deviation across epochs. Clean epochs (that belong to the same class/condition) should have a lower std dev across time than noise epochs. Alternatively, for datasets with 2 classes (e.g. standard P3 oddball) we could compute t-values across time... cleaner data should have higher t's.

jona-sassenhagen commented:
How about: rate of false positives (significant t-tests in baseline) to false negatives (significant t-tests for post-stimulus) estimated via bootstraping cleaned data sets?






On Tue, Jun 2, 2015 at 12:51 AM -0700, "Marijn van Vliet" <notifications@github.com> wrote:












I propose the following metric: we extract epochs, perform FASTER, then compute for each time sample, the standard deviation across epochs. Clean epochs (that belong to the same class/condition) should have a lower std dev across time than noise epochs. Alternatively, for datasets with 2 classes (e.g. standard P3 oddball) we could compute t-values across time... cleaner data should have higher t's.



—
Reply to this email directly or view it on GitHub.

wmvanvliet commented:
@jona-sassenhagen if all the data have multiple classes, that makes sense. What would the bootstrapping look like? Why not basic t-tests?

jona-sassenhagen commented:
I mean, bootstrap sets of datasets that will then be tested. Just so we have multiple t- test results for ttests across subjects. 



    _____________________________
From: Marijn van Vliet <notifications@github.com>
Sent: Dienstag, Juni 2, 2015 10:03 vorm.
Subject: Re: [mne-python] WIP: Automatic artifact detection with FASTER (#1777)
To: mne-tools/mne-python <mne-python@noreply.github.com>
Cc: jona-sassenhagen <jona.sassenhagen@gmail.com>


    

@jona-sassenhagen if all the data have multiple classes, that makes sense. What would the bootstrapping look like? Why not basic t-tests?   

—
Reply to this email directly or view it on GitHub.

wmvanvliet commented:
lets continue this in hangout

jona-sassenhagen commented:
Someone invite me, I'm on a train :)






On Tue, Jun 2, 2015 at 1:14 AM -0700, "Marijn van Vliet" <notifications@github.com> wrote:












lets continue this in hangout



—
Reply to this email directly or view it on GitHub.

dengemann commented:
Yes, I was gonna suggest that ;)

> On 02 Jun 2015, at 10:14, Marijn van Vliet <notifications@github.com> wrote:
> 
> lets continue this in hangout
> 
> —
> Reply to this email directly or view it on GitHub.
>

mainakjas commented:
okay, I am jumping on to the discussion. It seems that @wmvanvliet has been asking how to simulate the data. Was that solved? Don't we have functions for simulating dipoles? I think it's currently implemented for `evoked` and it should be possible to modify it for `epochs` too, no? Or am I missing something?

agramfort commented:
we have something for simultated data but it needs a lot of love...
usability and API is not great.

mainakjas commented:
okay. I was thinking about it only for the unit tests, not for method validation.

Between, I don't quite like the fact that all the methods start with `faster_`. I would go with a generic name like `find_bad_channels` and then add a `method` argument if better techniques come along. Maybe a function in `preprocessing` called `faster_pipeline` is fine with me, which internally calls all the 5 functions. But just once, not 5 times.

mainakjas commented on examples/preprocessing/plot_artifact_correction_with_faster.py:
```diff
@@ -0,0 +1,98 @@
+evoked_before = epochs_before.average()
+
+###############################################################################
+# Clean the data using FASTER
```
I vote for wrapping this entire pipeline into a single function

mainakjas commented on examples/preprocessing/plot_artifact_correction_with_faster.py:
```diff
@@ -0,0 +1,98 @@
+ica.apply(epochs)
+
+# Step 4: mark bad channels for each epoch and interpolate them. This is
+#         currently only implemented for EEG
```
now for MEG too thanks to yours truly ;)

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+from operator import add
+from scipy.signal import welch, lfilter
+from scipy.stats import kurtosis
+from functools import reduce
```
nitpick but I'd move this above the `scipy` imports

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+    if use_metrics is None:
+        use_metrics = metrics.keys()
+
+    # Concatenate epochs in time
```
this is also done in `decoding.classifier.ConcatenateChannels`. Maybe we need to refactor the common components ...

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+    if return_by_metric:
+        return bads
+    else:
+        return np.unique(np.concatenate(list(bads.values()))).tolist()
```
do we need a log somewhere like the epochs drop log which keeps track of the reason a channel was marked bad?

dengemann replied:
would be good.

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+    epochs : Instance of Epochs
+        The epochs for which bad channels need to be marked
+    picks : list of int | None
+        Channels to operate on. Defaults to EEG channels.
```
why only EEG ?

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+        The epochs for which bad channels need to be marked
+    picks : list of int | None
+        Channels to operate on. Defaults to EEG channels.
+    thres : float
```
minor nitpick, I'd name it `thresh` as that sounds like the natural place to break the word.

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+    max_iter : int
+        The maximum number of iterations performed during outlier detection
+        (defaults to 2, original FASTER paper uses 1).
+    return_by_metric : bool
```
do we really need this param? Just return a flat list and update the log?

dengemann replied:
It's useful for research.

mainakjas replied:
yeah, but updating the log would contain the same info ... see what I mean?

dengemann replied:
No I don't I. I suggested Marijn to add this because I wanted if for plotting and further processing (putting it in a dataframe, etc.).

mainakjas replied:
The drop log would contain what your dictionary contains right now.

dengemann replied:
I want the indices metric-wise as lists/ arrays, not to parse a log file

> On 06 Jul 2015, at 18:49, mainakjas <notifications@github.com> wrote:
> 
> In mne/preprocessing/faster.py:
> 
> > +    ----------
> > +    epochs : Instance of Epochs
> > +        The epochs for which bad channels need to be marked
> > +    picks : list of int | None
> > +        Channels to operate on. Defaults to EEG channels.
> > +    thres : float
> > +        The threshold value, in standard deviations, to apply. A channel
> > +        crossing this threshold value is marked as bad. Defaults to 3.
> > +    use_metrics : list of str
> > +        List of metrics to use. Can be any combination of:
> > +            'variance', 'correlation', 'hurst', 'kurtosis', 'line_noise'
> > +        Defaults to all of them.
> > +    max_iter : int
> > +        The maximum number of iterations performed during outlier detection
> > +        (defaults to 2, original FASTER paper uses 1).
> > +    return_by_metric : bool
> The drop log would contain what your dictionary contains right now.
> 
> —
> Reply to this email directly or view it on GitHub.
>

mainakjas replied:
okay. Let's keep it then. I can't think of a better solution now :)

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+        is presumed to be time.
+    sfreq : float
+        The sample rate of the timeseries.
+    prange : pair of floats
```
Is it a `tuple`? not quite clear. If it's a tuple, I think you should just say `tuple`, no?

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+    -------
+    fs : 1D numpy array
+        The frequencies for which the power spectra was calculated.
+    ps : ND numpy array
```
inconsistent documentation style

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+
+    Returns
+    -------
+    fs : 1D numpy array
```
you need to give the shape if it's 3D or less

mainakjas replied:
actually, I think this should be called `freqs`. Take a look at the TFR code.

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+
+def _efficient_welch(data, sfreq):
+    """Calls scipy.signal.welch with parameters optimized for greatest speed
+    at the expense of precision. The window is set to ~10 seconds and windows
```
why ~10 and not exactly 10? Can we be more precise?

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+        The epochs to analyze.
+    picks : list of int | None
+        Channels to operate on. Defaults to EEG channels.
+    thres : float
```
same comment here. `thres` -> `thresh`. Actually, looking at the current API, I am more and more convinced that we need just one function

dengemann replied:
What do you mean?

mainakjas replied:
I mean a single function that calls all the FASTER steps in sequence. I've described it elsewhere. The arguments to the functions look very similar to me. No point in asking the user to write 5 lines when it can be done in one line.

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+    max_iter : int
+        The maximum number of iterations performed during outlier detection
+        (defaults to 2, original FASTER paper uses 1).
+    return_by_metric : bool
```
same comments

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+        Defaults to all of them.
+    max_iter : int
+        The maximum number of iterations performed during outlier detection
+        (defaults to 2, original FASTER paper uses 1).
```
so we're not implementing vanilla FASTER as default? how much difference does it make? do we really need this parameter? If yes, it's not clear to the users where the iterations come in. We should redirect them to documentation of `find_outliers`

jona-sassenhagen replied:
I think it'd be sufficient if all of the tools were there as individual methods, and an example on the examples page showed how to do the full FASTER routine.

mainakjas replied:
So we don't need a `max_iter` param? Are you fine with my proposal to have individual methods (without the `faster_` prefix) and a wrapper function which just calls them in sequence?

jona-sassenhagen replied:
I think that would work with how I am currently using these tools, yes.

agramfort replied:
looks like we have a team for pair programming during the sprint ;)

dengemann replied:
I'd like to avoid commitments to internal outlier functions. It's been a mistake by the way to make bads.find_outliers public. We should first of all use the original plain vanilla FASTER to evaluate it.

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+    bads : list of str
+        The names of the bad EEG channels.
+    """
+    metrics = {
```
looks like there's a lot of duplication. But I may be wrong ...

wmvanvliet replied:
there's a little duplication... but not a lot and the code is pretty straightforward now. At each stage of FASTER, the input to the detection methods is slightly different. Sometimes it is a 1D array (single channel), sometimes a 2D array (single epoch), etc. Merging some code paths would make for slightly shorter, but more difficult to understand code.

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+        if hasattr(ica, '_pre_whitener') and ica.noise_cov is not None:
+            transform_matrix = np.dot(transform_matrix, ica._pre_whitener)
+
+    metrics = {
```
you should just define the dict somewhere and use the `update` method if there are minor changes

wmvanvliet replied:
each FASTER step would have some minor changes. It would make for more complicated code without much gain in brevity, so a net loss in maintainability of the code.

mainakjas replied:
ok. Can you address as much of the comments as you can, add tests, make sure the tests pass and bring this to a mergeable state?

dengemann replied:
I've got local changes here. Please wait for me. I addressed a severe bug with one of the funtions.

wmvanvliet replied:
how's it coming with that PR @dengemann ? It doesn't have to be pretty, I can clean it up on my end.

dengemann replied:
coming now, sorry had no time.

dengemann replied:
Marijn, you need to rebase, otherwise I cannot make a reviewable PR. I
already rebased locally.

2015-07-09 9:43 GMT+02:00 Marijn van Vliet <notifications@github.com>:

> In mne/preprocessing/faster.py
> <https://github.com/mne-tools/mne-python/pull/1777#discussion_r34230411>:
>
> > +    """
> > +    if power_gradient_range is None:
> > +        power_gradient_range = (25., 40.)
> > +
> > +    epochs = epochs.pick_channels(ica.info['ch_names'])
> > +    source_data = ica.get_sources(epochs).get_data().transpose(1, 0, 2)
> > +    source_data = source_data.reshape(source_data.shape[0], -1)
> > +
> > +    # Compute the transform matrix used by the ICA operator is necessary
> > +    if use_metrics is None or 'kurtosis' in use_metrics:
> > +        transform_matrix = np.dot(ica.mixing_matrix_.T,
> > +                                  ica.pca_components_[:ica.n_components_])
> > +        if hasattr(ica, '_pre_whitener') and ica.noise_cov is not None:
> > +            transform_matrix = np.dot(transform_matrix, ica._pre_whitener)
> > +
> > +    metrics = {
>
> how's it coming with that PR @dengemann <https://github.com/dengemann> ?
> It doesn't have to be pretty, I can clean it up on my end.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1777/files#r34230411>.
>

wmvanvliet replied:
done!

mainakjas commented:
ok done from my end. We need unit tests @wmvanvliet . Any chance you'll have time to work on it? We could start with what we have in the `simulation` module. And I think, the code needs some refactoring + API changes. I've made my suggestions. Thanks for the contribution !

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+            logger.info('Bad channel-in-epoch detection on %s channels:'
+                        % ch_type.upper())
+            s_epochs = metrics[m](data[:, chs])
+            for i, s in enumerate(s_epochs):
```
single letter variable names ...

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+            for i, s in enumerate(s_epochs):
+                b = [epochs.ch_names[picks[j]]
+                     for j in find_outliers(s, thres, max_iter)]
+                if len(b) > 0:
```
same here ... single letter variable names. Notes for @dengemann if he's making a new PR

mainakjas commented on mne/preprocessing/faster.py:
```diff
@@ -0,0 +1,463 @@
+
+    if picks is None:
+        picks = pick_types(epochs.info, meg=True, eeg=True, exclude=[])
+    if use_metrics is None:
```
we should do some checking to ensure that 'use_metrics' contains valid keys and throw a human-readable error message if not

wmvanvliet commented:
New, possibly relevant paper was just published:

http://doi.org/10.3389/fninf.2015.00016

mainakjas commented:
On Fri, Jul 10, 2015 at 6:32 PM, Marijn van Vliet <notifications@github.com>
wrote:

> New, possibly relevant paper was just published:
>
> http://doi.org/10.3389/fninf.2015.00016
>
fancy name again ;)

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1777#issuecomment-120453143>
> .
>

wmvanvliet commented:
If I were to design a method, I would call it CINDERELLA :)

dengemann commented:
HARDER, FASTER, BETTER, STRONGER ...

> On 10 Jul 2015, at 22:29, Marijn van Vliet <notifications@github.com> wrote:
> 
> If I were to design a method, I would call it CINDERELLA :)
> 
> —
> Reply to this email directly or view it on GitHub.
>

dengemann commented on mne/preprocessing/bads/bads.py:
```diff
@@ -0,0 +1,278 @@
+    if return_by_metric:
+        return bads
+    else:
+        return np.unique(np.concatenate(bads.values())).tolist()
```
@fraimondo cc look here, a new API that makes us independent of the method.

fraimondo replied:
Looks good. Now we need to develop the method we were discussing and plug
it there.
Le 13 juil. 2015 10:56 AM, "Denis A. Engemann" <notifications@github.com> a
écrit :

> In mne/preprocessing/bads/bads.py
> <https://github.com/mne-tools/mne-python/pull/1777#discussion_r34464565>:
>
> > +        picks = pick_types(epochs.info, meg=True, eeg=True, exclude=[])
> > +
> > +    if method == 'faster':
> > +        _method_params = dict(max_iter=1, thresh=3, use_metrics=None)
> > +        if isinstance(method_params, dict):
> > +            _method_params.update(method_params)
> > +        method_params = _method_params
> > +        bads = _faster._find_bad_channels(epochs, picks, **_method_params)
> > +    else:
> > +        raise NotImplementedError(
> > +            'Come back later, for now there is only "FASTER"')
> > +
> > +    if return_by_metric:
> > +        return bads
> > +    else:
> > +        return np.unique(np.concatenate(bads.values())).tolist()
>
> @fraimondo <https://github.com/fraimondo> cc look here, a new API that
> makes us independent of the method.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1777/files#r34464565>.
>

wmvanvliet commented:
I was unable to get the reference FASTER implementation to run and am giving up on it.

wmvanvliet commented:
After much deliberation, we've decided to drop the `find_bad_components` step. It never worked quite right and similar functionality is already exposed by the ICA code. The ICA code that deals with this could use some love, but that can be addressed in a separate PR.

@dengemann Lets merge this thing!

jona-sassenhagen commented:
I've accidentally rm'd my mne installation with my faster copy which left me wondering, what's the status here? Is a merge in sight?

dengemann commented:
@jona-sassenhagen can you take a look at the API and tell us what you think? the methods and ingredients will change, I don't think we will end up using faster.

jona-sassenhagen commented:
I've been using it with basically this API for now - I guess it works for me, although it becomes a bit long and maybe it could be considered merging some of the detection and application stages (e.g., allowing the function to detect bad epochs to return epochs objects with some epochs dropped).

Also: http://www.ncbi.nlm.nih.gov/pubmed/25791012
Paper on evaluating various methods for detecting bad ICs, including CORRMAP and FASTER. MATLAB code is on github: https://github.com/dnacombo/SASICA
I'll probably go about translating some of these into MNE code eventually.

kingjr commented:
Would you guys need testing/help at some point? @gmeade and I have a dataset of ~20 healthy subjects (60 EEG + 306 Neuromag) that could benefit from this code.

wmvanvliet commented:
@kingjr We would certainly appreciate help in evaluating the methods. Our main struggle at the moment is finding a test metric that makes sense. Datasets with manually flagged bad channels/epochs would be a great help.

dengemann commented:
I see it a bit differently. We have clear metrics and data. We don't have the right algorithms and are still thinking about a good API. FASTER will certianly not stay for long time.

jona-sassenhagen commented:
@wmvanvliet can you make a small commit so the routine is at least mergeable as-is? I'd like to deploy it on a few machines ...

wmvanvliet commented:
rebased.

jona-sassenhagen commented:
Thanks @wmvanvliet , you're a really fast coder!

Eric89GXL commented:
@wmvanvliet @dengemann do we have some conclusion about the status here?

dengemann commented:
I'm prone to freeze or even clone this PR. We need to do some science first
to find a principled way to deal with this. I don't see any progress in
sight here. My 2 cents.

On Thu, Sep 3, 2015 at 5:03 PM, Eric Larson <notifications@github.com>
wrote:

> @wmvanvliet <https://github.com/wmvanvliet> @dengemann
> <https://github.com/dengemann> do we have some conclusion about the
> status here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1777#issuecomment-137477959>
> .
>

wmvanvliet commented:
Closed. For Science.

kingjr commented:
Could you rebase and make travis happy so that I could try to give it a look?

kingjr commented:
Is there anyone continuing working on this?...

Shouldn't we separately merge some functions, and warn the user about their non-robustness? I would personally enjoy having `find_bad_channels` ...

agramfort commented:
@jasmainak's phd evolved in this direction... but don't tell anyone :)

dengemann commented:
And we decided that the latest API is good but is meaningless if we are not
convinced of any method.
Come back in one year or so :)

On Fri, Jan 22, 2016 at 9:36 AM, Alexandre Gramfort <
notifications@github.com> wrote:

> @jasmainak's phd evolved in this direction... but don't tell anyone :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1777#issuecomment-173844621>
> .
>

jasmainak commented:
> Come back in one year or so :)

lol, maybe that not long ;)

wmvanvliet commented:
Rebased this so you can check this out again. I may have found a neat way to evaluate this...

wmvanvliet commented:
How about this for a method of evaluation:

The underlying assumption of 'clean' EEG data is that there is some smoothness at the source level. If a single channel shows some peaks, but the surrounding channels do not, this is usually an indication of a bad channel.

So here is one way to evaluate the signal quality of EEG:

 1. Project EEG to source level
 2. Reconstruct EEG from source level using forward model
 3. Evaluate reconstruction error

By tweaking the lambda parameter, you can enforce some smoothness at the source level.

I made a gist that demonstrates the approach on the MNE sample data. It adds several types of noise and looks at the reconstruction error:

https://gist.github.com/wmvanvliet/dd17c9d8a9b75435217d

types of simulated noise:

 - **line_noise:** 50 Hz sine wave added to channel 12
 - **white_noise:** Gaussian noise added to channel 12
 - **filtered_noise:** Gaussian noise, lowpass filtered with 40Hz added to channel 12
 - **blink:** Simulated blink at 300ms by convolving a Gaussion PDF shape (std=0.1) with the EOG SSP projection
 - **clean:** the original, clean, data
![figure_1](https://cloud.githubusercontent.com/assets/428273/13638150/a5136e5c-e613-11e5-9401-0a402385a536.png)

You can see this method picks out 'bad' channels nicely, but fails to catch the blink noise, since the blink also has smoothness at the source level.

jasmainak commented:
That's indeed one way to go about it. I wish real data was as well-behaved as simulated data :)

wmvanvliet commented:
> I wish real data was as well-behaved as simulated data :)

Now I'm curious about the approach(es) you are trying!

kingjr commented:
I'm not sure I'm rediscovering the wheel, but I wrote this today, and it seems to work, at least on synthetic and personal data:

https://gist.github.com/kingjr/be5f8451b138eebd027b

The general idea is to try to predict the activity of each channel, from all other sensors, and measure the prediction error to identify bad channels. If the channel is artefacted, the error should be high. It's a bit similar to the mean correlation step of FASTER, but improved (notably because you don't have to specify the layout, and it works for MEG signals too, where the spatial correlation isn't always true).

Once you have identified the channel-specific noise, you can use @alexandrebarachant's pyRiemann Potato to identify the bad times (eye movements, muscle etc, saturation, line noise).

[I think the channel-specific vs time-specific noises should be distinguished, because they don't lead to the same correction (interpolate, versus reject).]

Let me know what you think

[Also, one of the alternative approaches that I thought of, but failed to implement consisted in finding the times and channels that minimize the variance of all channels, with a very strong regularization parameter to simultaneously minimize the number of bad times and bad channels. I tried that, but because the feature space can be huges, and the weights are boolean, I failed (I think the problem may not be strictly convex). @agramfort any thoughts in this direction?]

kingjr commented:
But maybe this is ~ equivalent to:
```
cov = compute_raw_covariance(raw, method='shrunk')
error_chan = np.zeros(n_chan)
for ch, weight in enumerate(cov.data):
    sel = np.delete(range(n_chan), ch)
    y_pred = np.dot(weight[sel], raw._data[sel, :])
    error_chan[ch] = np.mean((y_pred - raw._data[ch, :]) ** 2)
```
@dengemann ?

wmvanvliet commented:
MEG 0113 is obviously bad in the sample data. Can your method detect it?

wmvanvliet commented:
My approach is a way to obtain an overall score for the cleanliness of the signal. So it can be used as a verification that any bad channel detection algorithm (including your approach) does its job properly.

Here for example is a comparison between the original sample data (bads), the manually labeled bads removed (manual_bads) and any bads found by FASTER removed (faster_bads):

![figure_2](https://cloud.githubusercontent.com/assets/428273/13667074/73d77c70-e6bd-11e5-9cb9-5857c73c524d.png)

so from this toy example, the toy conclusion would be that FASTER seems to be doing a good job.

dengemann commented:
I think I said it earlier somewhere, we should focus here on a flexible API and then support options that have been published already. Then people can create their preproc workflows based on the preferred method. If we add new methods that might work we have to sit together and validate them on many datasets. If we don't want to do this use other tested method. The API would then segment the entire process into detecting bad channels, bad epochs, bad components, etc. and would come with method and method parameter parameters. Our problem at that point was that were not sure to which default strategies we wanted to get committed to. As long as we don't set a default we can of course add method by method with this architecture. With @fraimondo we actually extended and finished this API that we had proposed here, but using it for experimental detections strategies. I'd be ok with putting the faster if we don't make it a default or if we clearly communicate that being default does not mean being the best. Then we could put the API here and populate it with known options, e.g. PREP, Adjust, FASTEr, HCP and what have you. I think we should just focus on stuff that is used. I'm not so enthusiastic on putting research code here unless it is evaluated properly (= paper).

wmvanvliet commented:
So how do you evaluate with method (PREP, Adjust, FASTEr, HCP) is best?

jasmainak commented:
All of these are great ideas @kingjr and @wmvanvliet . For benchmarking, my feeling is that testing on real data over many (hundreds of) subjects is the right way to go about it. That way, you really test the limits of the different techniques.

kingjr commented:
> MEG 2443 and EEG 053 are obviously bad in the sample data. Can your method detect it?

Yes. here are the results using a single fit across all chan types from the entire non preprocessed raw (~35 s): The y axis is the reconstruction error, the x axis is irrelevant (channels)
![bad_chans](https://cloud.githubusercontent.com/assets/4881164/13672383/88f64520-e6a2-11e5-826b-8f968e2d17c5.png)

You can improve this by applying it to the epochs (tend to have more clean data).

I also noticed than it can detect channels that are bad during brief time periods. (e.g. if you take the mean reconstruction error across time, instead of the median, you become sensitive to rare but large errors).

Overall, we should probably break down the current issue into
* 1. providing simple examples and methods to detect bad channels (The examples/tutorials show how to interpolate them but not to detect them...)
* 2. deciding whether we incorporate some of the published steps of the FASTER, ADJUST etc monsters.
* 3. documenting which method we believe (based on published evidence and/or practice) is optimal  to limit the garden of forking paths.

and, (although I agree with @dengemann that the MNE rep is not the place for that)
* 4. developing new methods
* 5. validating previous methods on large multi-site multi-model already annotated datasets.

Eric89GXL commented:
> I think I said it earlier somewhere, we should focus here on a flexible API and then support options that have been published already. Then people can create their preproc workflows based on the preferred method.

An alternative two-step approach would be to:

1. Implement each approach as closely and completely as possible related to its paper implementation. Leave it nested e.g. in `mne.preprecessing`, and make it private.
2. Create a public API that gives access to multiple preprocessing methods once approaches are vetted and we have a sense of what options are shared across methods.

This helps decouple the implementation, API, and testing/use from one another a bit. That way we can add method-by-method as they get published, start comparing them and using them, yet not be bogged down by maintenance in the long run for methods that don't actually get used.

It's not a pretty solution, but I fear if we wait for all validation to be done before incorporating these methods in mne-python, we risk delaying their validation and use, which is a worse trade-off in my mind.

kingjr commented:
> It's not a pretty solution, but I fear if we wait for all validation to be done before incorporating these methods in mne-python, we risk delaying their validation and use, which is a worse trade-off in my mind.

+1000

I'm +1 for private

dengemann commented:
> An alternative two-step approach would be to:

this is exactly the API I described. With @wmvanvliet and @fraimondo we wrote a public API for exactly that. The idea is to choose from validated = published methods via string argument.

Eric89GXL commented:
On Mar 10, 2016 11:48 AM, "Denis A. Engemann" <notifications@github.com>
wrote:
>>
>> An alternative two-step approach would be to:
>
> this is exactly the API I described. With @wmvanvliet and @fraimondo we
wrote a public API for exactly that. The idea is to choose from validated =
published methods via string argument.

You mean my point 2 is what you described? I'm more saying that perhaps
FASTER or other methods should be merged in some form, independent of such
a public facing API, and then later incorporated into a monolithic function
if possible.

dengemann commented:
> m more saying that perhaps
FASTER or other methods should be merged in some form, independent of such
a public facing API, and then later incorporated into a monolithic function
if possible.

Yes yes yes. The idea is to have a common place to request artifact processing and choose between different "backends".

wmvanvliet commented:
Moved this to the sandbox:
https://github.com/mne-tools/mne-sandbox/pull/12
